core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -136,6 +146,16 @@ public RecoveryResponse recoverToTarget() throws IOException {
                 }
             }
 
+            // engine was just started at the end of phase 1
+            if (shard.state() == IndexShardState.RELOCATED) {
+                /**
+                 * we could have missed replication operations on the target if its engine wasn't started yet after primary relocation
```

Comment:
maybe, "the primary shard has been relocated while we copied files. This means we can't guarantee any more that any operation that was replicated during the file copy (when the target engine was not yet opened) will be present in the local translog and thus will be resent on phase 2. The reason is that operation replicated by the target primary are sent to recovery target and the local shard (old primary) concurrently, meaning it may have arrived at the recovery target before we opened the engine and is still on inlight on the local shard.

Checking the relocated status here, after we opened the engine on the target,  is safe because primary relocation waits for all ongoing operations to complete and be fully replicated. Therefore all future operation by the new primary are guaranteed to reach the target shard when it's engine is open.
"